pyo3-ffi/README.md

```diff
@@ -0,0 +1,188 @@
+        ml_flags: METH_FASTCALL,
+        ml_doc: "returns the sum of two integers as a string".as_ptr() as *const c_char,
+    };
+
+    PyModule_AddObject(
+        mptr,
+        "sum_as_string\0".as_ptr() as *const c_char,
+        PyCFunction_NewEx(
+            Box::into_raw(Box::new(wrapped_sum_as_string)),
+            std::ptr::null_mut(),
+            PyUnicode_InternFromString("string_sum\0".as_ptr() as *const c_char),
+        ),
+    );
+
+    let all = ["__all__\0", "__version__\0", "sum_as_string\0"];
+
+    let pyall = PyTuple_New(all.len() as isize);
+    for (i, obj) in all.iter().enumerate() {
+        PyTuple_SET_ITEM(
+            pyall,
+            i as isize,
+            PyUnicode_InternFromString(obj.as_ptr() as *const c_char),
+        )
+    }
+
+    PyModule_AddObject(mptr, "__all__\0".as_ptr() as *const c_char, pyall);
+
+    mptr
+}
+
+pub unsafe extern "C" fn sum_as_string(
+    _self: *mut PyObject,
+    args: *mut *mut PyObject,
+    nargs: Py_ssize_t,
+) -> *mut PyObject {
+    if nargs != 2 {
+        return raise_type_error("sum_as_string() expected 2 positional arguments");
+    }
+
+    let arg1 = *args;
+    if PyLong_Check(arg1) == 0 {
+        return raise_type_error("sum_as_string() expected an int for positional argument 1");
+    }
+    let arg1 = PyLong_AsLong(arg1);
+
+    let arg2 = *args.add(1);
+    if PyLong_Check(arg2) == 0 {
+        return raise_type_error("sum_as_string() expected an int for positional argument 2");
+    }
+
+    let arg2 = PyLong_AsLong(arg2);
+
+    let res = (arg1 + arg2).to_string();
+    PyUnicode_FromStringAndSize(res.as_ptr() as *const c_char, res.len() as isize)
+}
+
+#[cold]
+#[inline(never)]
+fn raise_type_error(msg: &str) -> *mut PyObject {
+    unsafe {
+        let err_msg =
+            PyUnicode_FromStringAndSize(msg.as_ptr() as *const c_char, msg.len() as isize);
+        PyErr_SetObject(PyExc_TypeError, err_msg);
+        Py_DECREF(err_msg);
+    };
+    std::ptr::null_mut()
+}
+```
+
+With those two files in place, now `maturin` needs to be installed. This can be done using
+Python's package manager `pip`. First, load up a new Python `virtualenv`, and install `maturin`
+into it:
+```bash
+$ cd string_sum
+$ python -m venv .env
+$ source .env/bin/activate
+$ pip install maturin
+```
+
+Now build and execute the module:
+```bash
+$ maturin develop
+# lots of progress output as maturin runs the compilation...
+$ python
+>>> import string_sum
+>>> string_sum.sum_as_string(5, 20)
+'25'
+```
+
+As well as with `maturin`, it is possible to build using [setuptools-rust] or
+[manually][manual_builds]. Both offer more flexibility than `maturin` but require further
+configuration.
+
+
+While most projects use the save wrapper provided by pyo3,
+you can take a look at the [`orjson`] as an example on how to use ffi directly.
+For those well versed in C and Rust the [tutorials] from the cpython documentation
+can be easily converted to rust aswell.
```

Comment:
Thanks for including this, this is a great example :)

Reply:
Its where I took a lot of inspiration from when I was working on the project which motivated this PR :) It's indeed very helpful

Reply:
I just noticed - it should be "safe" instead of "save" 😉

Reply:
... and "aswell" is not a word.